File: smart-hub/src/main/scala/io/github/tassiLuca/hub/infrastructure/ui/TemperatureSourceUI.scala
package io.github.tassiLuca.hub.infrastructure.ui

import java.awt.{BorderLayout, Component, Dimension}
import javax.swing.*

class TemperatureSourceUI(publishHandler: (String, Double) => Unit) extends JFrame:

  import io.github.tassiLuca.utils.ScalaSwingFacade.{*, given}

  private var sensors = 0

  setTitle("A simulated temperature source in a GUI pretending to be real")
  setPreferredSize(Dimension(500, 300))
  setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE)
  private val addButton = JButton("Add Sensor")
  private val middle = createPanel()
  middle.setLayout(BoxLayout(middle, BoxLayout.Y_AXIS))
  addButton.addActionListener(_ => addSensorTo(middle))
  private val mainPanel = createPanel(
    (addButton, BorderLayout.NORTH),
    (middle, BorderLayout.CENTER),
  )(using BorderLayout())
  getContentPane.add(mainPanel)

  private def addSensorTo(panel: JPanel): Component =
    sensors = sensors + 1
    val sendButton = JButton("Send")
    val sensorName = s"sensor-$sensors"
    val sensorInput = JTextField(4)
    sendButton.addActionListener(_ => publishHandler(sensorName, sensorInput.getText.toDouble))
    panel.addWithRepaint(createPanel(JLabel(sensorName), sensorInput, sendButton))

File: smart-hub/src/main/kotlin/io/github/tassiLuca/hubkt/adapters/ui/TemperatureSourceUI.kt
package io.github.tassiLuca.hubkt.adapters.ui

import java.awt.BorderLayout
import java.awt.Component
import java.awt.Dimension
import javax.swing.BoxLayout
import javax.swing.JButton
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTextField
import javax.swing.WindowConstants

/** A simulated temperature source in a GUI. */
class TemperatureSourceUI(private val publishHandler: (Pair<String, Double>) -> Unit) : JFrame() {

    private var sensors = 0

    init {
        title = "A simulated temperature source in a GUI pretending to be real"
        preferredSize = Dimension(WIDTH, HEIGHT)
        defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
        val middle = JPanel()
        middle.layout = BoxLayout(middle, BoxLayout.Y_AXIS)
        val addButton = JButton("Add Sensor")
        addButton.addActionListener { addSensorTo(middle) }
        val mainPanel = JPanel(BorderLayout()).apply {
            add(middle, BorderLayout.CENTER)
            add(addButton, BorderLayout.SOUTH)
        }
        contentPane.add(mainPanel)
    }

    private fun addSensorTo(panel: JPanel): Component {
        sensors++
        val sendButton = JButton("Send")
        val sensorName = "sensor-$sensors"
        val sensorInput = JTextField(4)
        sendButton.addActionListener { publishHandler(Pair(sensorName, sensorInput.text.toDouble())) }
        JPanel().apply {
            add(JLabel(sensorName))
            add(sensorInput)
            add(sendButton)
        }.also { panel.add(it) }
        panel.revalidate()
        panel.repaint()
        return panel
    }

    companion object {
        private const val WIDTH = 500
        private const val HEIGHT = 300
    }
}
